src/annotator/util/buckets.js

```diff
@@ -88,161 +95,151 @@ export function findClosestOffscreenAnchor(anchors, direction) {
+    if (current) {
+      buckets.push({
+        anchors: current.anchors,
+        position: current.position,
+      });
     }
+    if (!anchorPosition) {
+      return current;
+    }
+    const anchorHeight = anchorPosition.bottom - anchorPosition.top;
+    const bucketPosition = anchorPosition.top + anchorHeight / 2;
+    const currentBucket = /** @type WorkingBucket */ ({
+      anchors: [anchorPosition.anchor],
+      top: anchorPosition.top,
+      bottom: anchorPosition.bottom,
+      position: bucketPosition,
+    });
+    return currentBucket;
+  }
 
-    // For each point, we'll either:
-    // * create a new bucket: Add a new bucket (w/corresponding bucket position)
-    //   and add the working anchors to the new bucket. This, of course, has
-    //   the effect of making the buckets collection larger. OR:
-    // * merge buckets: In most cases, merge the anchors from the last bucket
-    //   into the penultimate (previous) bucket and remove the last bucket (and
-    //   its corresponding `bucketPosition` entry). Also add the working anchors
-    //   to the previous bucket. Note that this decreases the size of the
-    //   buckets collection.
-    // The ultimate set of buckets is defined by the pattern of creating and
-    // merging/removing buckets as we iterate over points.
-    const isFirstOrLastPoint =
-      buckets.length === 0 || index === points.length - 1;
+  // Build buckets from position information
+  anchorPositions.forEach(anchorPosition => {
+    if (anchorPosition.top < BUCKET_TOP_THRESHOLD) {
+      aboveScreen.add(anchorPosition.anchor);
+      return;
+    } else if (
+      anchorPosition.top >
+      window.innerHeight - BUCKET_BOTTOM_THRESHOLD
+    ) {
+      belowScreen.add(anchorPosition.anchor);
+      return;
+    }
 
+    if (!currentBucket) {
+      // We've encountered our first on-screen anchor position:
+      // We'll need a bucket!
+      currentBucket = _nextBucket(buckets, null, anchorPosition);
+      return;
+    }
+    // We want to contain overlapping highlights and those near each other
+    // within a shared bucket
+    const isContainedWithin =
+      anchorPosition.top > currentBucket.top &&
+      anchorPosition.bottom < currentBucket.bottom;
+
+    // The new anchor's position is far enough below the bottom of the current
+    // bucket to justify starting a new bucket
     const isLargeGap =
-      buckets.length &&
-      position - buckets[buckets.length - 1].position > BUCKET_GAP_SIZE;
-
-    if (current.anchors.length === 0 || isFirstOrLastPoint || isLargeGap) {
-      // Create a new bucket, because:
-      // - There are no more open/working anchors, OR
-      // - This is the first or last point, OR
-      // - There's been a large dimensional gap since the last bucket's position
-      buckets.push({ anchors: current.anchors.slice(), position });
+      anchorPosition.top - currentBucket.bottom > BUCKET_GAP_SIZE;
+
+    if (isLargeGap && !isContainedWithin) {
```

Comment:
So, basically, this seems to be the crux of the entire algorithm as I understand it. It seems quite simple, but seems to do the proverbial trick so far as I can ascertain.